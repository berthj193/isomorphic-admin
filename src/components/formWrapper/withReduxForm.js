import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';

export default reduxFormOptions => CustomComponent => {
  const ComponentWithReduxForm = reduxForm(reduxFormOptions)(CustomComponent);
  return class ReduxFormWrapper extends Component {
    static propTypes = {
      resetOnSuccess: PropTypes.bool,
      onSubmitSuccess: PropTypes.func,
    };

    handleSubmitSuccess = (result, dispatch, props) => {
      // props parameter contain properties of ComponentWithReduxForm from above! Keep that in mind!
      if (this.props.resetOnSuccess) {
        props.reset();
      }
      if (this.props.onSubmitSuccess) {
        this.props.onSubmitSuccess(result, dispatch, props);
      }
    };

    render() {
      const props = {
        ...this.props,
        onSubmitSuccess: this.handleSubmitSuccess,
      };

      return <ComponentWithReduxForm {...props} />;
    }
  };
};
