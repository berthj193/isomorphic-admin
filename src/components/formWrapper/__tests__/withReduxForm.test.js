import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { reduxForm } from 'redux-form';

import withReduxForm from '../withReduxForm';

const defaultOptions = {
  form: 'test-form',
};

const mockComponent = jest.fn().mockImplementation(() => <span />);

const getComponent = options => withReduxForm(options)(mockComponent);

jest.mock('redux-form', () => ({
  reduxForm: jest.fn().mockImplementation(() => Component => Component),
}));

describe('withReduxForm()', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render deliveredComponent', () => {
    const Component = getComponent(defaultOptions);
    expect(toJson(mount(<Component />))).toMatchSnapshot();
  });

  it('should initialize reduxForm with given options', () => {
    const options = {
      ...defaultOptions,
      myCustom: 'option',
    };

    getComponent(options);
    expect(reduxForm).toHaveBeenCalledWith(options);
  });

  it('should initialize reduxForm with given options', () => {
    const options = {
      ...defaultOptions,
      myCustom: 'option',
    };

    getComponent(options);
    expect(reduxForm).toHaveBeenCalledWith(options);
  });

  it('should run reset method from props if resetOnSuccess flag is set', () => {
    const options = {
      ...defaultOptions,
      myCustom: 'option',
    };

    const resetMock = jest.fn();

    const Component = getComponent(options);
    mount(<Component resetOnSuccess />).instance().handleSubmitSuccess(null, null, { reset: resetMock });
    expect(resetMock).toHaveBeenCalled();
  });

  it('should not run reset method from props if resetOnSuccess flag is unset', () => {
    const options = {
      ...defaultOptions,
      myCustom: 'option',
    };

    const resetMock = jest.fn();

    const Component = getComponent(options);
    mount(<Component />).instance().handleSubmitSuccess(null, null, { reset: resetMock });
    expect(resetMock).not.toHaveBeenCalled();
  });

  it('should run onSubmitSuccess method from props', () => {
    const options = {
      ...defaultOptions,
      myCustom: 'option',
    };

    const onSubmitSuccessMock = jest.fn();

    const Component = getComponent(options);
    mount(<Component onSubmitSuccess={onSubmitSuccessMock} />)
      .instance()
      .handleSubmitSuccess(null, null, { reset: () => null });
    expect(onSubmitSuccessMock).toHaveBeenCalled();
  });
});
