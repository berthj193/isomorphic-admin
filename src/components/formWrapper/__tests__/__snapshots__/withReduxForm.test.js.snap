// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withReduxForm() should render deliveredComponent 1`] = `
<ReduxFormWrapper>
  <mockConstructor
    onSubmitSuccess={[Function]}
  >
    <span />
  </mockConstructor>
</ReduxFormWrapper>
`;
